Chore(order): endpoints - Part 1:N

carlos-r-l-rodrigues opened the pull request:
- `archive` and `completeOrder` methods, workflows and endpoint
- Return Reasons endpoints.
- Incomplete skeleton of multiple missing endpoints with TODOs.

olivermrbl commented on packages/core/core-flows/src/order/steps/archive-orders.ts:
```diff
@@ -0,0 +1,39 @@
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+type CompleteOrdersStepInput = {
+  order_ids: string[]
```
nit(non-blocking): camelCase

olivermrbl commented on packages/core/core-flows/src/order/steps/complete-orders.ts:
```diff
@@ -0,0 +1,39 @@
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+type CompleteOrdersStepInput = {
+  order_ids: string[]
```
camelCase

olivermrbl commented on packages/core/core-flows/src/order/workflows/archive-orders.ts:
```diff
@@ -0,0 +1,17 @@
+import { archiveOrdersStep } from "../steps"
+
+type ArchiveOrdersStepInput = {
+  order_ids: string[]
```
ibid.

olivermrbl commented on packages/core/core-flows/src/order/workflows/complete-orders.ts:
```diff
@@ -0,0 +1,20 @@
+import { completeOrdersStep } from "../steps"
+
+type CompleteOrdersStepInput = {
+  order_ids: string[]
```
ibid.

olivermrbl commented on packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts:
```diff
@@ -0,0 +1,19 @@
+import { remoteQueryObjectFromString } from "@medusajs/utils"
+import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+
+export const GET = async (req: MedusaRequest, res: MedusaResponse) => {
```
thought(non-blocking): Wonder if we need this? Shouldn't we use /admin/payments/:id/refund instead?

carlos-r-l-rodrigues replied:
I prefer to use payments directly. Makes more sense to me.
I will remove it.

olivermrbl commented on packages/medusa/src/api-v2/admin/orders/[id]/cancel/route.ts:
```diff
@@ -0,0 +1,25 @@
+  MedusaResponse,
+} from "../../../../../types/routing"
+
+export const GET = async (
```
comment for future: Should be `POST`

olivermrbl commented on packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts:
```diff
@@ -0,0 +1,19 @@
+import { remoteQueryObjectFromString } from "@medusajs/utils"
+import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+
+export const GET = async (req: MedusaRequest, res: MedusaResponse) => {
```
comment for future: Should be `POST`

olivermrbl commented on packages/medusa/src/api-v2/admin/orders/fulfillments/[fulfillment_id]/cancel/route.ts:
```diff
@@ -0,0 +1,25 @@
+  MedusaResponse,
+} from "../../../../../../types/routing"
+
+export const POST = async (
```
thought(non-blocking): Wonder if we need this? We could have `/admin/fulfillments/:id/cancel` instead?

olivermrbl commented on packages/medusa/src/api-v2/admin/orders/fulfillments/route.ts:
```diff
@@ -0,0 +1,25 @@
+  MedusaResponse,
+} from "../../../../types/routing"
+
+export const POST = async (
```
thought(non-blocking): Wonder if we need this? Why not use `/admin/fulfillments` instead?

carlos-r-l-rodrigues replied:
You believe the fulfillments workflow could also update necessary things in an Order?

olivermrbl replied:
Ah yeah, you are right. There is probably some order-specific steps. Let's keep it 👍

olivermrbl commented on packages/medusa/src/api-v2/admin/returns/[id]/cancel/route.ts:
```diff
@@ -0,0 +1,35 @@
+  MedusaResponse,
+} from "../../../../../types/routing"
+
+export const GET = async (
```
thought: Needs a todo?

olivermrbl commented on packages/medusa/src/api-v2/admin/returns/[id]/receive/route.ts:
```diff
@@ -0,0 +1,36 @@
+  MedusaResponse,
+} from "../../../../../types/routing"
+
+export const GET = async (
```
questions: Needs a todo?

olivermrbl commented on packages/medusa/src/api-v2/admin/returns/[id]/route.ts:
```diff
@@ -0,0 +1,35 @@
+  MedusaResponse,
+} from "../../../../types/routing"
+
+export const GET = async (
```
ibid.

olivermrbl commented on packages/medusa/src/api-v2/admin/returns/route.ts:
```diff
@@ -0,0 +1,56 @@
+  const remoteQuery = req.scope.resolve(ContainerRegistrationKeys.REMOTE_QUERY)
+
+  const queryObject = remoteQueryObjectFromString({
+    entryPoint: "order",
```
question: Should be `return`. Also, do we need a todo here?

olivermrbl commented on packages/medusa/src/api-v2/store/return/validators.ts:
```diff
@@ -28,5 +29,12 @@ const ItemSchema = z.object({
+  return_shipping: ReturnShippingSchema,
+  note: z.string().optional(),
+  receive_now: z.boolean().optional(),
+  refund_amount: z.number().optional(),
```
todo: You shouldn't be able to pass `refund_amount` from the storefront

olivermrbl commented on packages/modules/order/src/services/order-module-service.ts:
```diff
@@ -2577,4 +2579,111 @@ export default class OrderModuleService<
+    if (notAllowed.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Pending Order cannot be archived: ${notAllowed.join(", ")}.`
```
```suggestion
        `Orders ${notAllowed.join(", ")} are completed, canceled, or in draft and cannot be archived`
```

olivermrbl commented on packages/modules/order/src/services/order-module-service.ts:
```diff
@@ -2577,4 +2579,111 @@ export default class OrderModuleService<
+    if (notAllowed.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Canceled Order cannot be completed: ${notAllowed.join(", ")}.`
```
```suggestion
        `Orders ${notAllowed.join(", ")} are canceled and cannot be completed`
```

olivermrbl reviewed (approved):
Overall LGTM, left a few comments here and there. Still some work left, but this will allows us to finish the UI-side of orders, which is nice.

To not miss these todos, maybe we should create tickets in Linear.

carlos-r-l-rodrigues commented:
@olivermrbl I am addressing all your comments in a following PR.
Thanks a loto for the review, there are a good chunk of flows left in the Order domain that require us changing a much broader scope than just the Order module, so these details are left behind.
